Add Code Repository Search Agent

pranath-reddy opened the pull request:
### Summary :memo:
This PR adds an initial implementation of the `CodeSearchAgent`, a configurable agent that performs iterative code repository search by reformulating queries. It coordinates multiple sub agents and tools to enable a reasoning-driven search. It takes a user query, reformulates it using `QueryAgent`, performs code search via a pluggable `CodeSearchTool`, evaluates results using a rubric-based `MultiRubricRelevancyAgent`, and refines queries through a `FollowUpQueryAgent`. The process continues iteratively until sufficient alignment is reached or the iteration limit is hit.

### Details
- Introduced `CodeSearchAgent` under `akd.agents.search.code_search` to support iterative, LLM-guided code repository search.
- Inherits from `ControlledSearchAgent` to leverage existing search logic and iteration control.
- Fully modular via dependency injection:
  - `QueryAgent` for reformulating the initial query.
  - `CodeSearchTool` (e.g., `CombinedCodeSearchTool`, `LocalRepoCodeSearchTool`) for retrieving search results.
  - `MultiRubricRelevancyAgent` for evaluating alignment of search results with query intent.
  - `FollowUpQueryAgent` for refining queries based on content and feedback.
- Extends `ControlledSearchAgentConfig` with code-specific parameters:
  - Embedding model configuration for local search
  - SDE search API configuration
  - Tool selection flags for local/SDE search components
  - Model selection for sub-agents
- Added `is_server_available` under akd.utils for URL format validation and reachability testing

### Usage
```python
from akd.agents.search import CodeSearchAgent, CodeSearchAgentConfig
from akd.agents.search import LitSearchAgentInputSchema

# Configure and initialize the agent
cfg = CodeSearchAgentConfig()
agent = CodeSearchAgent(config=cfg)

# Run the search
result = await agent.arun(
    LitSearchAgentInputSchema(
        query="landslide nepal", 
        max_results=10
    )
)
```
```python
from akd.utils import is_server_available

url = "https://www.google.com"
is_server_available(url)
```

### Checks
- [x] Tested Changes
- [x] Stakeholder Approval

pranath-reddy committed:
Add Code Search Agent

- Add initial version of code search agent
- Inherits from Controlled Search agent
- Also added tests and examples

NISH1001 commented on akd/agents/search/code_search.py:
```diff
@@ -0,0 +1,187 @@
+
+    # SDE search configuration
+    sde_base_url: str = Field(
+        default="https://d2kqty7z3q8ugg.cloudfront.net/api/code/search",
```
can we by default get from os env `SDE_BASE_URL` and default to this URL?

NISH1001 commented on akd/agents/search/code_search.py:
```diff
@@ -0,0 +1,187 @@
+        default="https://d2kqty7z3q8ugg.cloudfront.net/api/code/search",
+        description="SDE search API base URL",
+    )
+    sde_search_type: str = Field(
```
use `Literal["vector", "hybrid", "keyword"]` to restrict the values

NISH1001 commented on akd/agents/search/code_search.py:
```diff
@@ -0,0 +1,187 @@
+
+        self.config = config or CodeSearchAgentConfig()
+        if debug is not None:
+            self.config.debug = debug
```
Is this required?

NISH1001 commented on akd/agents/search/code_search.py:
```diff
@@ -0,0 +1,187 @@
+        query_agent: Optional[QueryAgent] = None,
+        followup_query_agent: Optional[FollowUpQueryAgent] = None,
+        relevancy_agent: Optional[MultiRubricRelevancyAgent] = None,
+        debug: Optional[bool] = None,
```
`debug: bool = False` is better way

NISH1001 commented on akd/agents/search/code_search.py:
```diff
@@ -0,0 +1,187 @@
+        self,
+        config: Optional[CodeSearchAgentConfig] = None,
+        search_tool: Optional[CombinedCodeSearchTool] = None,
+        query_agent: Optional[QueryAgent] = None,
```
Let's use new python 3.12+ type hint style
```python
config: CodeSearchAgentConifg | None = None
search_tool: CombinedSearchTool | None = None
...
...
```

NISH1001 commented on akd/agents/search/code_search.py:
```diff
@@ -0,0 +1,187 @@
+    def __init__(
+        self,
+        config: Optional[CodeSearchAgentConfig] = None,
+        search_tool: Optional[CombinedCodeSearchTool] = None,
```
The type for this should be `CodeSearchTool` no? Which wil give same semantics no matter what input tools are.

NISH1001 commented on akd/agents/search/code_search.py:
```diff
@@ -0,0 +1,187 @@
+                page_size=self.config.sde_page_size,
+            )
+            sde_tool = SDECodeSearchTool(config=sde_config)
+            tools.append(sde_tool)
```
Do we need this?

I was thinking, since the incoming `search_tool` will get passed something or None.
```python
search_tool = search_tool or CombinedSearchTool([LocalTool(...), SDE...()])
```

This way, if we want to enable what to have for SDE etc, we can depednecny-inject direclty.

Maybe, for this method, better way could be to have a classmethod-initilized thing:

```python
class CodeSearchAgent(...):
..
...

@classmethod
def from_local_sde_config(cls, config_agent: CodeSearchAgentConfig | None = None, local_search_config...., sde_search_config,....) -> CodeSearchAgent:
  # initialize combined tool here direclty and pass to the constructor
  tools = []
  if local_search_config:
    tools.append(Local...(local_search_config))...
  elif sde_search_config:
    ....
  tool = CombinedSearchTool(...)
  return cls(config=agent_config, tool=tool, ...)
```

Something like this. This iwll help to get more ctonrol on how to initialize and segregate responsibility

pranath-reddy replied:
I am creating the tools internally since I didn't know whether having external tool configs/definitions would be compatible with the node template.

pranath-reddy committed:
PR comment updates-1

- Addressed the minor PR comments
To-Do:
- Update the search tool instantiation

pranath-reddy committed:
Update search tool instantiation

- Prioritize explicitly provided search_tool over defaults
- Add checks for local data file and SDE URL reachability
- Log warnings when tools are unavailable

NISH1001 commented on akd/agents/search/code_search.py:
```diff
@@ -0,0 +1,248 @@
+        if search_tool is not None:
+            self.search_tool = search_tool
+        else:
+            self.search_tool = self._setup_search_tool()
```
```python
self.search_tool = search_tool or self._setup_search_tool()
```

NISH1001 commented on akd/agents/search/code_search.py:
```diff
@@ -0,0 +1,248 @@
+    )
+
+
+class CodeSearchAgent:
```
Let's inherit from `ControlledSearchAgent` and rework some logic accordingly as most of the logic we can reuse from the parent class. ANd only change the logic if needed?

NISH1001 commented on akd/agents/search/code_search.py:
```diff
@@ -0,0 +1,248 @@
+        self.relevancy_agent = relevancy_agent or self._setup_relevancy_agent()
+
+        # Create the underlying ControlledSearchAgent
+        self._controlled_agent = ControlledSearchAgent(
```
if we inherit from `ControlledSEarchAgent`, we don't need to hae this as a sub agent as the whole agent will be this agent.

NISH1001 reviewed (changes requested):
Pass I

pranath-reddy committed:
Update code_search.py

- Updated code search agent to inherit from ControlledSearchAgent

NISH1001 commented on akd/agents/search/code_search.py:
```diff
@@ -0,0 +1,227 @@
+    embedding_model_name: str = Field(
+        default="thenlper/gte-large", description="Embedding model for local search"
+    )
+    data_file: str | None = Field(
+        default=None, description="Path to local repository data file"
+    )
```
What data file is this?

pranath-reddy replied:
It's for the repo dump

NISH1001 commented on akd/agents/search/code_search.py:
```diff
@@ -0,0 +1,227 @@
+                    if self.config.debug:
+                        logger.warning(
+                            f"[CodeSearchAgent] Could not reach SDE URL '{url}'. Skipping SDE tool. Reason: {e}"
+                        )
```
Can we maybe move this test to `akd.utils.py`? That is: maybe `is_server_available(url...) -> bool` or something and reuse that? It could be handy for some other cases also within akd.

```python
if self.config.use_sde_search:
  reachable = is_server_available(sde_base_url)
```

```python
def is_server_available(url: str | HttpURL) -> bool:
  # test for url validity first
  # then requests.head(...)
  # if any exceptions or non 2xx status, return False...
```

pranath-reddy replied:
okay, will add to utils

NISH1001 commented on akd/agents/search/code_search.py:
```diff
@@ -0,0 +1,227 @@
+                if self.config.debug:
+                    logger.warning(
+                        f"[CodeSearchAgent] SDECodeSearchTool unavailable; continuing without it. Reason: {e}"
+                    )
```
If we have the `is_server_available` util, then i think this method will look less complicated and help flatten the nested try catch as well.

pranath-reddy committed:
Merge remote-tracking branch 'origin/develop' into code-search-agent

pranath-reddy commented:
> @pranath-reddy could you pull in the latest develop branch. There, the failed tests are 0 so far. So need to reflect the test coverage here as well. Thanks.

One failed test. Looks like there's a missing field in the output of code search agent. Will fix it. Was LitSearchAgentOutputSchema updated?

NISH1001 commented on akd/agents/search/code_search.py:
```diff
@@ -0,0 +1,227 @@
+                if self.config.debug:
+                    logger.warning(
+                        f"[CodeSearchAgent] LocalRepoCodeSearchTool unavailable; continuing without it. Reason: {e}"
+                    )
```
also maybe segregate the logic to an ewm ethod like `_setup_local_tool(...) -> LocalSearchTool | None` or something?

This iwll clean the setup method here?

pranath-reddy committed:
Fix code search agent

- Fixed an issue where LitSearchOutputSchema requires query to be added to the result item which was causing the test to fail.

pranath-reddy committed:
Add new util

- Added new util is_server_available to check if a url is reachable

pranath-reddy committed:
Update code_search.py

- segregated sub tools creation to separate methods
